tensorflow/java/src/main/java/org/tensorflow/op/core/Zeros.java

```diff
@@ -0,0 +1,56 @@
+package org.tensorflow.op.core;
+
+import java.nio.ByteBuffer;
+
+import org.tensorflow.DataType;
+import org.tensorflow.Operand;
+import org.tensorflow.Output;
+import org.tensorflow.Shape;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Scope;
+import org.tensorflow.op.annotation.Operator;
+
+/**
+ * An operator creating a constant initialized with zeros w.r.t its type and shape.
+ *
+ * @param <T> constant type
+ */
+@Operator
+public class Zeros<T> implements Op, Operand<T> {
```

Comment:
This is a bit different from say [`tf.zeros`](https://www.tensorflow.org/api_docs/python/tf/zeros) in Python, where the shape can be a tensor (i.e., something that isn't fully known at graph construction time).

For example:

```python
import tensorflow as tf

shape = tf.random_uniform([2], minval=1, maxval=5, dtype=tf.int32)
zeros = tf.zeros(shape)

with tf.Session() as sess:
  for _ in range(5): print sess.run(zeros.shape)
```

You'll see different shapes of the zeros on each call to `sess.run` - it isn't a graph construction time constant.
Wouldn't we want the same for the Java zeros?

Reply:
Would it make sense to split out the `Zeros.java` file out of this PR and then have a separate PR  that adds a fully functional `Zeros`? Or is there some more immediate need for graph-construction-time-constant-zeros that you have in mind?

Reply:
I meant that both `Zeros` and `ZerosLike` take (as input) tensors computed at graph execution time. The former uses the input as the shape of the output, the latter uses the shape of the input tensor as the shape of the output tensor. For example, in:

```python
import tensorflow as tf

x = tf.placeholder(tf.int32)
z0 = tf.zeros(x)
z1 = tf.zeros_like(x)

with tf.Session() as sess:
  print(sess.run([z0, z1], feed_dict={x: [1, 1]}))
```
`z0` will evaluate to `[[0.]]`, while `z1` will evaluate to `[0., 0.]`.
Note that `tf.zeros` in a [pretty trivial wrapper](https://github.com/tensorflow/tensorflow/blob/25c197e02393bd44f50079945409009dd4d434f8/tensorflow/python/ops/array_ops.py#L1550) over [`tf.fill`](https://www.tensorflow.org/api_docs/python/tf/fill).

Also, the shape is known only at graph execution time. And when it is known, it is fully specified.
Which is why I think having `Zeros.create()` use a `Shape` argument isn't quite right. The type of the `shape` argument should be `Output<Integer>` or `Output<Float>`.

Something like:

```java
public class Zeros<T> implements Op, Operand<T> {
  public static <T> Zeros<T> create(Scope scope, Class<T> type, Output<Integer> shape) {
    return create(scope, type, shape);
  }

  public static <T> Zeros<T> create(Scope scope, Class<T> type, Output<Long> shape) {
    return create(scope, type, shape);
  }

  private static <T, U> Zeros<T> create(Scope scope, Class<T> type, Output<U> shape) {
    Output<T> zero = Constant.create(scope, type, new long[]{}, ByteBuffer.allocate(DataType.fromClass(type).byteSize());
     return Fill<T>(scope, type, shape, zero);
   }
}    
```

And callers would do something like:

```java
Output<Float> zeroMatrix = ops.zeros(ops.constant(new long[]{2, 2}));
```

Let me know if that makes sense, or if I'm misunderstanding.

Reply:
Interesting... So when calling `tf.zeros([2, 2])` in python, it creates a shape tensor out of the array parameter? 

So to resume. without the `Zeros` operator, we could achieve the same operation doing

```ops.fill(Float.class, ops.constant(new long[]{2, 2}, Long.class), ops.constant(0.0f));```

And with `Zeros`, it would be

```ops.zeros(Float.class, ops.constant(new long[]{2, 2}, Long.class))```

It’s still very verbose, that is why I liked the way Python does it, by passing a simple array. Or at least not passing the `Long.class` here would make it more succinct... is it frequent to create shape tensors out of an array like this? If so, maybe we could have special factory in `Constant` (or another operator) that takes only a `long[]` as parameter?

Reply:
Yes, `tf.zeros([2, 2])` creates a shape tensor out of the array parameter. In fact, _all_ Python APIs create tensors where appropriate out of the arguments. For example, `tf.matmul([[1]], [[2]])` will result in the exact same graph as `tf.matmul(tf.constant([[1]]), tf.constant([[2]]))`, since the Python APIs convert the arguments to `tf.Tensor` objects wherever possible. To do the same in Java would be something like using `Object` for all arguments :):

```java
public Tensor<T> matmul(Object a, Object b) throws Exception {
  Output<T> aTensor = convertToOutputOrRaiseException(a);
  Output<T> bTensor = convertToOutputOrRaiseException(b);
  ...
}
```

So, in that sense `Zeros` is no more special than any of the other existing ops - `Add`, `MatMul` etc. 

Adding overrides in the `Constant` operator that parallel the ones in [`Tensors.create`](https://www.tensorflow.org/api_docs/java/reference/org/tensorflow/Tensors) seems fine to me.

Reply:
All right, let's to this then! And `ops.fill()` won't have the same overflow issue that I had with the previous implementation, so that's good